### ISSUE TITLE
Scapy import slow on Windows

### ISSUE THREAD

**Issue by rgom** [secdev/scapy#925]
When using scapy on Windows, there are the following problems:
1. Scapy import  ("from scapy import all as scapy") takes ages to complete. It's about 12+ seconds
2. Scapy import might hang when executed from IDE (Eclipe/PyDev)

The first one is caused by time spent on:
- powershell parsing (exec query)
- pcap parsing (pcapdnet)

Profiling revealed that the corresponding methods are executed several times with the same arguments.
By adding a poor man's caching to all of them I've reduced the time by half (to about 6 seconds).
Still not good, but quite impressive, IMO.
I don't know if caching can be always done there (adding/changing interfaces), but in my case it helps a lot.

3.5 s of that is spent in remaining powershell queries. Some of them could be possibly replaced with Python libraries (netifaces, psutil, ...).

For the second issue. The code hangs IIRC on subprocess.popen. Somebody else was able to reproduce it (it didn't occur on my machine), but to be safe I replaced read in exec_query with communicate, to avoid potential deadlocks. I'll have to check that.

Regards,
Robert

**Comment by gpotter2** [comment]
HI, thanks for your report

have you tried using the very latest dev version ?
https://github.com/secdev/scapy/archive/master.zip
We recently have improved quite a lot the loading time (to me about 4s average)

**Comment by rgom** [comment]
Thank you for the heads up.
I'm sorry for not mentioning the version - here it is:

```
MINGW64 /c/repos/git/scapy (master)
$ git describe
v2.3.3-862-g72cb6a3
```

And it's only 3 commits behind origin/master, so should be quite new:

```
MINGW64 /c/repos/git/scapy (master)
$ git status
On branch master
Your branch is behind 'origin/master' by 3 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

nothing to commit, working tree clean

MINGW64 /c/repos/git/scapy (master)
$ git remote -v
origin  https://github.com/secdev/scapy.git (fetch)
origin  https://github.com/secdev/scapy.git (push)
```

And the slowness ... see for yourself:

![image](https://user-images.githubusercontent.com/8234247/32137306-a2f851aa-bc1d-11e7-9723-e1cfb690e252.png)

The 3 huge blocks correspond to:
- exec query ps (6s)
- get_if_list (pcapdnet) (3s)
- get_if_raw_addr (pcapdnet) (3s)

As said before, caching is a low hanging fruit (only that there's not lru_cache in Python 2, so has to be implemented manually), but even after applying remaining powershell queries take 3.5 s :\

I can help with some of that, if needed.

Btw. how do you measure loading time? In my case it's:

```
import cProfile

cProfile.run("from scapy import all as scapy", "c:\\temp\\slow_scapy5")
```

Robert

**Comment by gpotter2** [comment]
You got the point, we definatly need to speed up the loading time...

Here are my ideas yet:
- in pcapdnet, use a cache system. Because it's calling winpcap API, we can't really speed the process
- in arch/windows, the powershell calls are slow because powershell take ages to start. We could get handle it so we would start a single powershell instance, and keep it during all startup.

We are looking for a no-dependency solution (scapy should be runnable on its own). 

To measure loading time, I'm simply using
```
a = time.time(); from scapy.all import *; print(time.time()-a)
```
I didn't know cProfiles, which looks extremely useful ^^

**Comment by rgom** [comment]
Thanks for the heads up.

I started digging a bit to attack the most severe issue first - PowerShell queries.
Other than caching, I was thinking about replacing that with something simpler, as I believe they introduce at least the following layers: subprocess, shell parsing.
As I see it, PowerShell queries could be replicated with win32com client queries. I used them, but while they alleviated the mentioned problems, they were somehow slower (numerous COM calls).
Then I found those links:
  - http://www.christophekeller.com/using-ctypes-to-call-additional-win32-functions/ (and then https://github.com/flying-circus/pywingui/tree/master/pywingui/network) - quite heavy, covering more aspects than needed
- https://stackoverflow.com/questions/3787825/problem-with-using-win32-iphelper-api-in-python - supposedly lightweight, no external dependencies, using ctypes

Today was a day of experiments, I'd have to perform proper performance measurements to see if there's a benefit in using it here. And it's small, it should be no problem in adding such a code to scapy, if no external dependencies are required.

Regards,
Robert

**Comment by gpotter2** [comment]
Thanks for your answer.

It could be a good way out to use win32 queries, and we can be sure it's fast.
However, it adds a quite complicated code, and might break in future windows releases. Moreover, current implentation allows a lot of flexibility, as we can parse any properties of the WMI classes.

I personally try to avoid using win32 requests in python, but i'll be pleased if they speed up the booting time.

I've added a small cache system on pcapdnet in https://github.com/secdev/scapy/pull/927. I also corrected a small bug retrieving the routes twice in certain cases, which you might have been affected by.

Another option would be to use the same powershell instance. Here's a proof of concept, which I'm not sure yet if it's faster: https://github.com/gpotter2/scapy/tree/win-speedup-2. Could you please try it and tell me if scapy boots faster ?

Thanks a lot !

Regards
Gabriel

**Comment by rgom** [comment]
Thanks!

It seems that you've hit the same issue as me - problems with communicating to subprocess:

```
>>> from scapy import all as scapy
PS C:\Users\myusername>
```

So the version from win-speedup-2 is not usable at this moment.

Regards,
Robert

**Comment by gpotter2** [comment]
My bad, the issue should be fixed by now (it works on my Windows 10 computer).
```
gpotter@surface MINGW64 ~/Documents/gpotter/scapy (win-speedup-2)
$ git status
On branch win-speedup-2
Your branch is up-to-date with 'origin/win-speedup-2'.
nothing to commit, working tree clean
```
Then
```
>>> from scapy.all import *
>>>
```
(Please note that i have used `python setup.py install` again to re-install scapy, before calling python)

**Comment by rgom** [comment]
I'm very sorry, but the last version from win-speedup-2 behaves the same -
hangs when used from IDE, enters powershell when called from console
(conemu+ipython or idle).

What I do (to be clear):
1. git checkout win-speedup-2 (in this case)
2. git fetch --all; git rebase
3. python setup.py bdist_wininst
4. install scapy from exe (after killing processes, if needed)
5. execute from console:
>>> from scapy.all import *
PS C:\Users\username>  <-- waits for powershell input
6. or execute from IDE (with help of -m trace --trace) - hangs, last lines
are:
__init__.py(85):         while True:
__init__.py(86):
self.buffer.append(self.process.stdout.readline())
threading.py(613):             if not self.__flag:
threading.py(615):             return self.__flag
 --- modulename: threading, funcname: __exit__
threading.py(289):         return self.__lock.__exit__(*args)
__init__.py(81):         for i in range(0,3):
__init__.py(82):             self.process.stdout.readline()
__init__.py(85):         while True:
__init__.py(86):
self.buffer.append(self.process.stdout.readline())
__init__.py(81):         for i in range(0,3):
__init__.py(82):             self.process.stdout.readline()
__init__.py(85):         while True:
__init__.py(86):
self.buffer.append(self.process.stdout.readline())

(So the lines are different than previously, indicating that some changes
were applied)

BTW. It's Windows 7 64 bit + Python 2.7.13 32 bit

Feel free to bother me if you need more info.

Regards,
Robert

**Comment by gpotter2** [comment]
@rgom Hello,

You are very right, the Windows 7 part still does not work. Windows 8+ and 7 are using different functions, and it seems that the windows 7 ones are still glitchy.

I'll investigate this

**Comment by guedou** [comment]
Can we close this issue ? It seems that it was fixed with #927

**Comment by rgom** [comment]
Generally, this problem seems to be fixed and the issue can be closed.
Remaining issues (if any) will be presumably reported separately.

Thanks,
Robert
